File: leetcode/matrix/74.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // // Method 1: O(m+n)
        // int m, n;
        // m = matrix.size();
        // n = matrix[0].size();
        // int i = 0;
        // while ( i < m && target >= matrix[i][0] ) {
        //     ++i;
        // }
        // if ( i > 0 ) {
        //     i -= 1;
        // }
        // for ( int j = 0; j < n; ++j ) {
        //     if ( matrix[i][j] == target ) {
        //         return 1;
        //     }
        // }
        // return 0;

        // Method 2: BInary search O(log(mn)) = O(log(m)+log(n))
        int m, n, left, right, mid, temp;
        m = matrix.size();      // col
        n = matrix[0].size();   // row
        left = 0;
        right = m*n - 1;

        while ( left <= right ) {
            mid = left + (right - left) / 2;
            temp = matrix[mid/n][mid%n];
            if ( temp == target ) {
                return true;
            } else if ( temp < target ) {
                left = mid + 1;
            } else if ( temp > target ) {
                right = mid - 1;
            }
        }
        return false;
    }
};

File: leetcode/binary search/74.ts
function searchMatrix(matrix: number[][], target: number): boolean {
  // // Method 1: 模数轮转+二分查找
  // const h = matrix.length;
  // const w = matrix[0].length;
  // const hw = h*w;

  // function bs(): boolean {
  //     let left = 0;
  //     let right = hw - 1;
  //     while ( left <= right ) {
  //         let mid = left + Math.floor((right - left)/2);
  //         let y = Math.floor(mid/w);
  //         let x = mid % w;
  //         if ( matrix[y][x] === target ) {
  //             return true;
  //         } else if ( matrix[y][x] > target ) {
  //             right = mid-1;
  //         } else if ( matrix[y][x] < target ) {
  //             left = mid+1;
  //         }
  //     }
  //     return false;
  // }

  // return bs();

  // Method 2: 两次二分查找，但需要边界查询
  const h = matrix.length;
  const w = matrix[0].length;

  let left = 0;
  let right = h - 1;
  while ( left <= right ) {
      let mid = left + Math.ceil((right - left)/2);
      if ( matrix[mid][0] === target ) {
          left = mid+1;
      } else if ( matrix[mid][0] > target ) {
          right = mid-1;
      } else if ( matrix[mid][0] < target ) {
          left = mid+1;
      }
  }
  if ( right < 0 ) {
    return false;
  }
  let resultY: number = right;

  left = 0;
  right = w - 1;
  while ( left <= right ) {
      let mid = left + Math.floor((right - left)/2);
      if ( matrix[resultY][mid] === target ) {
          return true;
      } else if ( matrix[resultY][mid] > target ) {
          right = mid-1;
      } else if ( matrix[resultY][mid] < target ) {
          left = mid+1;
      }
  }

  return false;
};